### PR TITLE
Merge Storage Blob Client Builders

### PR DESCRIPTION
**Pull request by alzimmermsft**
Merges the Storage Blob client builders into a single builder class which is able to instantiate all blob client types.

### CODE REVIEW

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java [hunk 1]
```diff
@@ -121,29 +129,77 @@ private AzureBlobStorageBuilder buildImpl() {
 
     /**
      * @return a {@link BlobClient} created from the configurations in this builder.
+     * @throws NullPointerException If {@code endpoint}, {@code containerName}, or {@code blobName} is {@code null}.
      */
-    public BlobClient buildClient() {
-        return new BlobClient(buildAsyncClient());
+    public BlobClient buildBlobClient() {
```
Comment (jaschrep-msft) [inline storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: Now that we build all four types of blob clients from the same builder, we should have the docstrings each have a little something of a main description talking about what the clients are for. Specifically, I want them to touch on how Blob(Async)Client is for generic blob methods like download and properties for when you don't know what kind of blob it is, and that the other three are for use when you know the type of blob you are interacting with.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java [hunk 2]
```diff
@@ -121,29 +129,77 @@ private AzureBlobStorageBuilder buildImpl() {
...
+    }
+
     /**
      * Sets the service endpoint, additionally parses it for information (SAS token, container name, blob name)
      * @param endpoint URL of the service
      * @return the updated BlobClientBuilder object
-     * @throws IllegalArgumentException If {@code endpoint} is a malformed URL.
+     * @throws IllegalArgumentException If {@code endpoint} is {@code null} or is a malformed URL.
      */
     public BlobClientBuilder endpoint(String endpoint) {
```
Comment (jaschrep-msft) [inline storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: This method needs to be able to accept a fully formed URL to a blob resource. This only takes in the scheme, host, snapshot, and SAS; the path is never looked at.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java [hunk 3]
```diff
@@ -158,35 +214,37 @@ public BlobClientBuilder endpoint(String endpoint) {
             if (parts.snapshot() != null) {
                 this.snapshot = parts.snapshot();
             }
```
Comment (jaschrep-msft) [inline storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: So I think this branch condition is a little off (once this method also uses the path of the URL). Take this set of steps:

1. `myClient1 = builder.endpoint("http://myaccount.blob.core.windows.net/mycontainer/myblob?snapshot=mysnapshotid").buildBlobClient();`
2. `myClient2 = builder.endpoint("http://myaccount.blob.core.windows.net/anothercontainer/anotherblob").buildBlobClient();`

Step 2 will NOT erase the snapshot in the builder, and myClient2 will point to `http://myaccount.blob.core.windows.net/anothercontainer/anotherblob?snapshot=mysnapshotid`, at best a nonexistent resource and at worst the wrong one.

We need to make decisions about what we overwrite under what circumstances when this method is used, as well as clearly document it. My first thoughts include:

1. Host and scheme are always overwritten
2. If any part of the path is specified, then we overwrite container name, blob name, and snapshot id. (if there is only one path segment, we assume it's the implicit root container) (we don't need to worry about only the snapshot being specified, as you can't snapshot an account)
3. SAS token is always overwritten, including setting it to null if not present

Input from @rickle-msft and @JonathanGiles is much appreciated.

File: storage/client/blob/src/test/java/com/azure/storage/blob/APISpec.groovy [hunk 4]
```diff
@@ -42,8 +40,6 @@ class APISpec extends Specification {
 
     static final ByteBuffer defaultData = ByteBuffer.wrap(defaultText.getBytes(StandardCharsets.UTF_8))
 
-    static final Flux<ByteBuffer> defaultFlux = Flux.just(defaultData)
```
Comment (jaschrep-msft) [inline storage/client/blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: Are the items removed in this file just not being used anywhere?

Comment (alzimmermsft) [reply storage/client/blob/src/test/java/com/azure/storage/blob/APISpec.groovy]: Correct, removed as nothing was referencing this.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java (hunk 3 above)
Comment (alzimmermsft) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: I agree with those proposals, they are not only easier to explain but also removes assumptions we are trying to make on the behalf of the consumer.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java (hunk 2 above)
Comment (alzimmermsft) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: Good catch, I'll clean that up.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java (hunk 1 above)
Comment (JonathanGiles) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: I agree!

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java (hunk 2 above)
Comment (alzimmermsft) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: Turns out that `URLParser` which this call into handles splitting the path into containerName and blobName.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java (hunk 1 above)
Comment (alzimmermsft) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: Added comments to explain when to build which client.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java (hunk 2 above)
Comment (alzimmermsft) [reply storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: Turns out the path is handled by the URLParser, it is turned into containerName and blobName pieces.

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java [hunk 5]
```diff
@@ -152,6 +168,10 @@ public AppendBlobClient buildAppendBlobClient() {
     }
 
     /**
+     * Creates a {@link AppendBlobAsyncClient} based on options set in the Builder. AppendBlobAsyncClients are used to
+     * perform append blob specific actions such as {@link AppendBlobAsyncClient#appendBlock(Flux, long)}, only use this
```
Comment (rickle-msft) [inline storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: Looks like it should be {@link AppendBlobAsyncClient#appendBlock(Flux, long) append block} 

The "append block" at the end is missing

File: storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java [hunk 6]
```diff
@@ -160,6 +180,10 @@ public AppendBlobAsyncClient buildAppendBlobAsyncClient() {
     }
 
     /**
+     * Creates a {@link BlockBlobClient} based on options set in the Builder. BlockBlobClients are used to perform block
```
Comment (rickle-msft) [inline storage/client/blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: Can you modify it to "perform generic upload operations such as uploadFromFile in addition to block blob specific operations..."

**Pull request merged**